Joe/pytorch checkpointing

visionjo opened the pull request:
Create a script to train and validate the model. Allow for models to train from scratch or start from the previous checkpoint, provided the checkpoint file exists.

visionjo committed:
Create interactive_commad_list

visionjo committed:
Create restore_from_checkpoint.py

visionjo committed:
Create submit.bash

visionjo committed:
Create train_with_checkpoints.py

visionjo committed:
Implemented training and validation 

save file out

visionjo committed:
Removed unused function

visionjo committed:
Added device var

visionjo committed:
Created function to save checkpoints

visionjo committed:
save checkpoints and check for best model via function call

visionjo committed:
Fix imports

visionjo committed:
Create function to load checkpoint

visionjo committed:
Reformat

visionjo committed:
make Exercise_4 python run script works end-to-end

visionjo committed:
Update interactive_command_list file

visionjo committed:
Cleanup code

manasvita commented:
@visionjo, just remember to pick 1-2 reviewers for this to start the PR process. Once approved, you can go ahead and merge it to main, until then it will remain blocked.

visionjo committed:
Update README

Correct and add info. For instance, Tutorial slides were listed as being in each subfolder, which they are not. Also, now that it is getting larger, summarizing
all contents, while specifying which exercise is whicj

finally, added a To Do section

visionjo committed:
Draft submit.bash for Exercise_4

visionjo committed:
Update interactive command list for Exercise_4

visionjo committed:
Add ToDos

visionjo committed:
Add To Do

visionjo commented:
`Exercise_4` was added for PyTorch. Use the interactive command list in the same directory to test.

Also, `README` in the main folder of the repo was updated, as it was out-of-date; plus, with more exercises, it seemed a more elaborate README was appropriate. This is a work-in-progress but included the barebones in PR.

manasvita commented on Exercise_4/interactive_commad_list:
```diff
@@ -0,0 +1,33 @@
+
+# Train for a single epoch
+train_with_checkpoints.py --help
+python restore_from_checkpoint.py --data /work/bootcamp/tutorials/ --epochs 2 --out "training_4" --checkpoint
```
restore_from_checkpoint.py is an empty file.

manasvita commented on Exercise_4/interactive_commad_list:
```diff
@@ -0,0 +1,33 @@
+cd Exercise_4
+
+# Train for a single epoch
+train_with_checkpoints.py --help
```
This command doesn't work and doing

python train_with_checkpoints.py --help 

results in syntax error.

manasvita commented on Exercise_4/submit.bash:
```diff
@@ -0,0 +1,22 @@
+#SBATCH --mem=10Gb		#Request 10GB of RAM memory
+#SBATCH --output=%A-%a.out      #set output filename with main job ID and task array ID
+#SBATCH --error=%A-%a.err       #set error filename with main job ID and task array ID
+#SBATCH --array=1-10%1		#execute 10 array jobs, 1 at a time.
```
Based on this current setup, it's not obvious why array job is required here. I have the same doubt about Exercise_2 as well.

manasvita reviewed (commented):
@visionjo, thanks for putting this together. I went through our current material on checkpointing, both on canvas & RTD, and I believe that it needs major revamp. 

For starters, it would be good to add 3 sub-sections to the checkpointing page:
1. User-level
2. Application-level
3. System-level

Each sub-page will then have its corresponding examples. The scripts in Exercise_1 are indeed for User-level but need modification in terms of documentation and some checking conditions. Currently, all those are missing.

Exercise_2 & Exercise_4 are examples of Application-level checkpointing. It would be good to point to the source pages, TF & PyTorch, and have some more in-depth explanations as to when and how the checkpointing is being carried out and the checkpointed model is being loaded. Both webpages explain this quite nicely, so we don't need to rewrite that but just point out bits relevant to our cluster so that it's easier for our users to follow. 
As for Slurm array jobs, I'm still not convinced that our current examples actually show how to execute checkpointing while utilizing Slurm's job array feature. So need to think about better ways.
Finally, should use this as an example to revamp our RTD on checkpointing and provide some more relevant examples, including how to utilize Slurm for this, which we haven't tested yet
https://www.cism.ucl.ac.be/Services/Formations/checkpointing.pdf

visionjo commented:
@manasvita, this one was a draft. Just so you know – it will be completed shortly. This will come after the others are completed.

visionjo commented:
Thanks for the notes! Duly noted, @manasvita, once I return to this!
